cmd/syncthing/gui.go

```diff
@@ -369,12 +390,23 @@ func (s *apiService) Serve() {
 		close(s.startedOnce)
 	}
 
-	// Serve in the background
+	// Serve in the background for each listener
 
 	serveError := make(chan error, 1)
-	go func() {
-		serveError <- srv.Serve(listener)
-	}()
+	for i, listener := range listeners {
+		srv := http.Server{
+			Handler: handlers[i],
+			// ReadTimeout must be longer than SyncthingController $scope.refresh
+			// interval to avoid HTTP keepalive/GUI refresh race.
+			ReadTimeout: 15 * time.Second,
+		}
+		go func(listener net.Listener) {
+			select {
+			case serveError <- srv.Serve(listener):
+			default:
```

Comment:
Why the select-default? AFAIK this means that there will never be sent anything to `serveError` as `srv.Serve` does not return immediately.

Reply:
Good catch.
What needs to happen is
```
err := srv.Serve(listener)
select {
case serverError <- err:
default:
}
```

Reply:
Actually it might still work, as I suspect srv.Serve needs to be evaluated.
Needs testing.

Reply:
I believe your right concerning your last point. Actually, I have wondered whether I really have to evaluate the `Serve()` call before the `select`. So, I had a look on the semantic specification of the `select` statement. Here it is (as picked up from https://golang.org/ref/spec#Select_statements)

 Execution of a "select" statement proceeds in several steps:

1. For all the cases in the statement, the channel operands of receive operations and the channel and right-hand-side expressions of send statements are evaluated exactly once, in source order, upon entering the "select" statement. The result is a set of channels to receive from or send to, and the corresponding values to send. Any side effects in that evaluation will occur irrespective of which (if any) communication operation is selected to proceed. Expressions on the left-hand side of a RecvStmt with a short variable declaration or assignment are not yet evaluated.
2. If one or more of the communications can proceed, a single one that can proceed is chosen via a uniform pseudo-random selection. Otherwise, if there is a default case, that case is chosen. If there is no default case, the "select" statement blocks until at least one of the communications can proceed.
3. Unless the selected case is the default case, the respective communication operation is executed.
4. If the selected case is a RecvStmt with a short variable declaration or an assignment, the left-hand side expressions are evaluated and the received value (or values) are assigned.
5. The statement list of the selected case is executed.

Reply:
Point taken, everything is correct as is. And probably according to typical Go-style (as terse as possible) this is the right way to do it. I would prefer Audrius variant as it is more obvious what happens, but then again I should just remember how select works.